flamedisx/likelihood.py

```diff
@@ -181,19 +169,17 @@ def set_data(self,
             elif dname not in self.dsetnames:
                 raise ValueError(f"Dataset name {dname} not known")
 
-        if autograph:
-            # When using traced ll, always retrace after setting data since
-            # n_events might be different
-            self.trace_log_likelihood()
-
     def simulate(self, rate_multipliers=None, fix_truth=None, **params):
         """Simulate events from sources, optionally pass custom
         rate_multipliers and fix_truth.
         """
         if rate_multipliers is None:
             rate_multipliers = dict()
 
-        ds = []
+        # Collect Source event DFs in ds, adding empty DataFrame ensures
+        # pd.concat doesn't fail if n_to_sim is 0 for all sources or all
+        # sources return 0 events
+        ds = [pd.DataFrame()]
```

Comment:
Maybe you can add this right before the concatenate, so the code below is easier to understand (`len(ds) > 2` -> `len(ds) > 1`, `ds[1:]` -> `ds`). Or just skip the concatenate if not len(ds).